File: app/lib/models/thread.dart
import 'package:kurl/models/kurl_result.dart';
import 'package:kurl/models/message.dart';
import 'package:kurl/models/user_ref.dart';

class MessagePreview {
  final String? body;
  final KurlResult? kurl;
  final String senderUid;
  final String createdAt;

  const MessagePreview({
    this.body,
    this.kurl,
    required this.senderUid,
    required this.createdAt,
  });

  factory MessagePreview.fromJson(Map<String, dynamic> json) {
    return MessagePreview(
      body: json['body'],
      kurl: json['kurl'] == null ? null : KurlResult.fromJson(json['kurl']),
      senderUid: json['senderUid'] ?? '',
      createdAt: json['createdAt'] ?? '',
    );
  }
}

class MessageThread {
  final String uid;
  final UserRef user; // the other participant
  final String lastMessageAt;
  final int unread;
  final MessagePreview? lastMessage; // thread-list rows only
  final String? lastReadAt; // thread-detail header only

  const MessageThread({
    required this.uid,
    required this.user,
    required this.lastMessageAt,
    this.unread = 0,
    this.lastMessage,
    this.lastReadAt,
  });

  factory MessageThread.fromJson(Map<String, dynamic> json) {
    return MessageThread(
      uid: json['uid'],
      user: UserRef.fromJson(json['user']),
      lastMessageAt: json['lastMessageAt'] ?? '',
      unread: json['unread'] ?? 0,
      lastMessage: json['lastMessage'] == null
          ? null
          : MessagePreview.fromJson(json['lastMessage']),
      lastReadAt: json['lastReadAt'],
    );
  }
}

class ThreadDetail {
  final MessageThread thread;
  final List<Message> messages;

  const ThreadDetail({required this.thread, required this.messages});

  factory ThreadDetail.fromJson(Map<String, dynamic> json) {
    return ThreadDetail(
      thread: MessageThread.fromJson(json['thread']),
      messages: (json['messages'] as List? ?? const [])
          .map((e) => Message.fromJson(e))
          .toList(),
    );
  }
}

File: api/src/models/thread.py
"""
Thread Model
Field mapping for the threads list row and for a single thread's header.
"""

from models.message import loads


def thread_summary(row) -> dict:
    return {
        "uid": row["uid"],
        "user": {"uid": row["other_uid"], "username": row["other_username"]},
        "lastMessageAt": row["last_message_at"],
        "unread": row["unread"] or 0,
        "lastMessage": _preview(row),
    }


def _preview(row) -> dict | None:
    if not row["last_at"]:
        return None
    return {
        "body": row["last_body"],
        "kurl": loads(row["last_kurl"]),
        "senderUid": row["last_sender_uid"],
        "createdAt": row["last_at"],
    }


def thread_header(row, other: dict, viewer_uid: str) -> dict:
    last_read = row["last_read_a_at"] if row["user_a_uid"] == viewer_uid else row["last_read_b_at"]
    return {
        "uid": row["uid"],
        "user": {"uid": other["uid"], "username": other["username"]},
        "lastMessageAt": row["last_message_at"],
        "lastReadAt": last_read,
        "createdAt": row["created_at"],
    }
